File: dynamic_programming/fibonacci_number.cpp
// Top-down implementation
class Solution {
    std::unordered_map<int, int> storage;
public:
    // TC: O( n )
    // SC: O( n )
    int fib(int n) {
        if( n == 0 )
            return 0;
        
        if( n == 1 )
            return 1;
        
        if( storage.count( n ) )
            return storage[ n ];
        
        storage[ n ] = fib( n - 1 ) + fib( n - 2 );
        return storage[ n ];
    }
};

// Bottom-Up implementaion
class Solution {
public:
    // TC: O( n )
    // SC: O( 1 )
    int fib(int n)
    {
        if( n <= 1 )
            return n;

        int first = 0;
        int second = 1;
        for( int i = 2; i <= n; ++i ) {
            int tmp = first + second;
            first = second;
            second = tmp;
        }
        
        return second;
    }
};

File: dynamic_programming/fibonacci_number.rs
// Bottom-Up implementation
impl Solution {

    // TC: O( n )
    // SC: O( 1 )
    pub fn fib(n: i32) -> i32 {
        if n <= 1 {
            return n;
        }
        
        let mut first = 0;
        let mut second = 1;
        for i in 2..n + 1 {
            let tmp = second + first;
            first = second;
            second = tmp;
        }
        second
    }
}

// Top-down implementatino
// TC: O( n )
// SC O( n )
impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n == 0 {
            return 0;
        }
        
        if n == 1 {
            return 1;
        }
        
        Self::fib( n - 1 ) + Self::fib( n - 2 )
    }
}
